### PR TITLE
Fix a bug in dual variables and reduced costs when we add implied bounds on free variables

### LINKED ISSUE #1119: [BUG] QP constraint duals returned by cuOpt are off by a factor of 2
## Summary

The Lagrange multiplier (constraint dual) returned by cuOpt for QP problems is off by a factor of two. Primal solution and primal objective are correct; only the dual is wrong. This was surfaced by the cvxpy nightly test `TestCUOPT::test_cuopt_qp_0` (`cvxpy/tests/test_conic_solvers.py`), which asserts the dual of an active inequality constraint and fails with:

```
AssertionError: 1.0000000092281611 != 2 within 4 places (0.9999999907718389 difference)
```

Reproduces with both `SolverMethod.PDLP` and `SolverMethod.Concurrent`. The factor-of-2 is consistent across distinct QPs (see case 2 below), so this is a scaling bug in the dual path, not a test-specific edge case.

## Reproducer

```python
import numpy as np
from cuopt.linear_programming.data_model import DataModel
from cuopt.linear_programming.solver import Solve
from cuopt.linear_programming.solver_settings import SolverSettings, SolverMethod
from cuopt.linear_programming.solver.solver_parameters import CUOPT_METHOD

# QP: min x^2  s.t.  x >= 1
# KKT:  gradient = 2x, at x*=1 -> dual = 2
# cuOpt convention: obj = c^T x + x^T Q x with Q symmetrized as Q+Q^T
# (data_model.py:286). To encode (1/2) x^T P x with P=2 (i.e. x^2),
# pass Q_input = P/2 = 1 (same convention cvxpy's cuopt_conif uses).
dm = DataModel()
dm.set_csr_constraint_matrix(
    np.array([1.0]), np.array([0], dtype=np.int32), np.array([0, 1], dtype=np.int32)
)
dm.set_constraint_lower_bounds(np.array([1.0]))
dm.set_constraint_upper_bounds(np.array([np.inf]))
dm.set_variable_lower_bounds(np.array([-np.inf]))
dm.set_variable_upper_bounds(np.array([np.inf]))
dm.set_objective_coefficients(np.array([0.0]))
dm.set_quadratic_objective_matrix(
    np.array([1.0]), np.array([0], dtype=np.int32), np.array([0, 1], dtype=np.int32)
)

ss = SolverSettings()
ss.set_parameter(CUOPT_METHOD, SolverMethod.PDLP)
res = Solve(dm, ss)
print("primal obj :", res.get_primal_objective())   # ~1.0   (correct)
print("primal sol :", res.get_primal_solution())    # ~[1.0] (correct)
print("dual sol   :", res.get_dual_solution())      # ~[1.0] <-- EXPECTED [2.0]
```

Output on cuOpt 26.06.00:
```
primal obj : 1.0000000186763234
primal sol : [1.00000001]
dual sol   : [1.00000001]    <-- should be ~[2.0]
```

### Case 2 (to rule out coincidence)

`min (x-3)^2 s.t. x >= 5`  =>  `x^2 - 6x + 9`, optimum x*=5, gradient at optimum = 2(x-3) = 4, so dual = 4.

Passing `c = [-6]`, `Q_input = [1]`:

```
primal obj : -4.99999998945835    (= 5^2 - 6*5, correct modulo ignored offset)
primal sol : [5.]                 (correct)
dual sol   : [2.]                 <-- should be [4]
```

Factor of 2 again. Note ignoring offset +9 here for simplicity; the point is the dual.

## Analysis

The C API (`cpp/include/cuopt/linear_programming/cuopt_c.h:258`) specifies the objective as

```
minimize  c^T x + x^T Q x + offset
```

with Q internally symmetrized as `Q + Q^T` (`python/.../data_model/data_model.py:286`). That implies the solver forms the effective quadratic as `(1/2) x^T (Q + Q^T) x` (otherwise the primal objective would also be off, but it isn't).

For the dual/reduced-cost computation in the PDLP QP path, the gradient of the objective should be `(Q + Q^T) x + c`. The observed factor-of-2 is consistent with using only `Q x + c` (missing the `+Q^T x` contribution) when computing the KKT residual that feeds the dual.

Entry points to look at:
- `cpp/src/pdlp/solve.cu` dual/reduced-cost post-processing
- `python/cuopt/cuopt/linear_programming/solver/solver_wrapper.pyx:374-375` (dual extraction — direct passthrough, no scaling)

## Impact

- Breaks cvxpy nightly `TestCUOPT::test_cuopt_qp_0`.
- Any cvxpy user relying on QP constraint duals from cuOpt (shadow prices, sensitivity analysis) gets values half the correct magnitude. Silent correctness bug — primal looks fine.

## Environment

- cuOpt 26.06.00 (main, commit c034027b at time of repro)
- Platform: Linux x86_64, Python 3.x
- Methods tested: `SolverMethod.PDLP`, `SolverMethod.Concurrent` — both wrong by the same factor

## Suggested next steps

1. Add a cuOpt Python test exercising QP constraint dual correctness (e.g. the two cases above — currently `python/cuopt/cuopt/tests/quadratic_programming/test_qp.py::test_solver` only checks primal).
2. Fix the QP dual/reduced-cost computation to use `(Q + Q^T) x` for the gradient term.
3. Confirm the fix unblocks cvxpy `test_cuopt_qp_0`.

### PR DESCRIPTION
**Pull request by chris-maes**
When -inf <= x_j <= inf and we add implied bounds on x_j, we will compute a dual solution (u, w) that satisfies

A^T u + w = c + Qx

but w_j could be nonzero. When x_j is free, we need to enforce that the corresponding reduced cost is zero. We do this by transforming (u, w) into (y, z) such that A^T y + z = c + Qx with z_j = 0 for all j such that -inf < x_j < inf.

We also add a unit test on a small QP to confirm the fix.

Fixes #1119

### CODE REVIEW

**Review by ramakrishnap-nv — APPROVED** [review]

File: cpp/src/dual_simplex/presolve.cpp [hunk 1]
```diff
@@ -1711,6 +1744,32 @@ void uncrush_solution(const presolve_info_t<i_t, f_t>& presolve_info,
...
+  // y_{i*} += du and z_k -= a_{i*,k} * du for all k in constraint i*.
+  if (!presolve_info.bounded_free_variables.empty()) {
+    settings.log.printf("Post-solve: Correcting duals for %d bounded free variables\n",
+                        static_cast<i_t>(presolve_info.bounded_free_variables.size()));
+    const csc_matrix_t<i_t, f_t>& A = original_problem.A;
+    for (const auto& bfv : presolve_info.bounded_free_variables) {
+      const f_t w_j = input_z[bfv.variable];
+      if (w_j == 0.0) { continue; }
+      const f_t du = w_j / bfv.coefficient;
+      input_y[bfv.constraint] += du;
```
[the remark this answers is not in the document]
Comment (chris-maes) [reply cpp/src/dual_simplex/presolve.cpp:1783]: Take a look at where bfv.coefficient comes from. Can it every be zero or non-zero?

**Review by rg20 — APPROVED** [review]

File: cpp/src/dual_simplex/presolve.cpp [hunk 2]
```diff
@@ -973,6 +988,24 @@ i_t presolve(const lp_problem_t<i_t, f_t>& original,
...
+        bounding_coefficient = lower_bound_coefficient[j];
+      } else if (problem.upper[j] < inf && upper_bound_constraint[j] != -1) {
+        bounding_constraint  = upper_bound_constraint[j];
+        bounding_coefficient = upper_bound_coefficient[j];
+      }
+      if (bounding_constraint != -1) {
+        presolve_info.bounded_free_variables.push_back(
+          {j, bounding_constraint, bounding_coefficient});
+      }
+    }
```
[the remark this answers is not in the document]
Comment (chris-maes) [reply cpp/src/dual_simplex/presolve.cpp:1007]: I don't understand what you are saying. In uncrush we revisit in reverse order. Please check yourself and if you still think it is an issue, please explain in more detail.

[the remark this answers is not in the document]
Comment (chris-maes) [reply cpp/src/dual_simplex/presolve.cpp:1007]: Ah yes. You could potentially make a z[x0] no longer zero. I fixed this.

File: cpp/tests/dual_simplex/unit_tests/solve_barrier.cu [hunk 3]
```diff
@@ -174,4 +181,39 @@ TEST(barrier, dual_variable_greater_than)
...
+
+  auto solution = cuopt::linear_programming::solve_lp(&handle, mps_data, settings);
+
+  EXPECT_EQ((int)solution.get_termination_status(), CUOPT_TERMINATION_STATUS_OPTIMAL);
+
+  auto h_x = cuopt::host_copy(solution.get_primal_solution(), handle.get_stream());
+  auto h_y = cuopt::host_copy(solution.get_dual_solution(), handle.get_stream());
+  auto h_z = cuopt::host_copy(solution.get_reduced_cost(), handle.get_stream());
+
+  printf("x %e y %e z %e\n", h_x[0], h_y[0], h_z[0]);
```
Comment (mlubin) [inline cpp/tests/dual_simplex/unit_tests/solve_barrier.cu:211]: printf still needed?

**Review by mlubin — APPROVED** [review]
Didn't double check the math.

File: cpp/tests/dual_simplex/unit_tests/solve_barrier.cu (hunk 3 above)
Comment (chris-maes) [reply cpp/tests/dual_simplex/unit_tests/solve_barrier.cu:211]: Still used to see results of tests
